grafana/data_source_folders.go

```diff
@@ -0,0 +1,95 @@
+package grafana
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"strings"
+
+	gapi "github.com/grafana/grafana-api-golang-client"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func DatasourceFolders() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: readFolders,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Description: `
+* [Official documentation](https://grafana.com/docs/grafana/latest/dashboards/dashboard-folders/)
+* [HTTP API](https://grafana.com/docs/grafana/latest/developers/http_api/folder/)
+`,
+
+		Schema: map[string]*schema.Schema{
+			"folders": {
+				Type:        schema.TypeSet,
+				Computed:    true,
+				Description: "The Grafana instance's folders.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"title": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The folder title.",
+						},
+						"id": {
+							Type:        schema.TypeInt,
+							Computed:    true,
+							Description: "The folder ID.",
+						},
+						"uid": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The folder's unique identifier.",
+						},
+						"url": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The folder's URL",
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func readFolders(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client).gapi
+	folders, err := client.Folders()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	uids := []string{}
+	for _, folder := range folders {
+		uids = append(uids, folder.UID)
+	}
+	sha := sha256.Sum256([]byte(strings.Join(uids, "-")))
+
+	d.SetId(fmt.Sprintf("%x", sha))
```

Comment:
The ID should be declarative because it needs to be unique. For a datasource like this, it's usually the combination of all entry parameters because you will typically not have the same datasource twice with the same parameters

With no parameters, you can just set the ID as `grafana_folders`. No need to calculate a SHA